pkg/controllers/webhookcerts/cert_requester.go

```diff
@@ -0,0 +1,422 @@
+// to the CSR API for a client certificate to be used by the kube-apiserver when
+// it's requesting token reviews from the oauth-apiserver
+func NewWebhookAuthenticatorCertRequester(
+	operatorClient v1helpers.OperatorClient,
+	csrClient certv1client.CertificateSigningRequestInterface,
+	csrInformers certv1informers.CertificateSigningRequestInformer,
+	secretClient corev1client.SecretInterface,
+	secretInformers corev1informers.SecretInformer,
+	eventsRecorder events.Recorder,
+) factory.Controller {
+	c := &webhookAuthenticatorCertRequester{
+		operatorClient: operatorClient,
+
+		csrClient: csrClient,
+		csrLister: csrInformers.Lister(),
+
+		secretClient: secretClient,
+		secretLister: secretInformers.Lister(),
+	}
+
+	return factory.New().
+		WithSync(c.syncWithConditions).
+		WithSyncDegradedOnError(operatorClient).
+		WithInformers(operatorClient.Informer()).
+		WithFilteredEventsInformers(nameFilter(csrName), csrInformers.Informer()).
+		WithFilteredEventsInformers(nameFilter(certSecretName), secretInformers.Informer()).
+		ToController("WebhookAuthenticatorCertRequester", eventsRecorder.WithComponentSuffix("webhook-authenticator-cert-requester"))
+}
+
+func (c *webhookAuthenticatorCertRequester) syncWithConditions(ctx context.Context, syncCtx factory.SyncContext) error {
+	progressing, err := c.sync(ctx, syncCtx)
+	if err != nil {
+		return err
+	}
+
+	// nil == keep the current progressing status
+	if progressing == nil {
+		return nil
+	}
+
+	var progressingText string
+	progressingStatus := operatorv1.ConditionFalse
+	progressingReason := "CertificateAvailable"
+	if *progressing {
+		progressingText = "waiting for CSR to be signed"
+		progressingStatus = operatorv1.ConditionTrue
+		progressingReason = "WaitingForCertificate"
+	}
+
+	_, _, statusErr := v1helpers.UpdateStatus(c.operatorClient,
+		v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
+			Type:    "AuthenticatorCSRProgressing",
+			Status:  progressingStatus,
+			Reason:  progressingReason,
+			Message: progressingText,
+		}))
+	return statusErr
+}
+
+func (c *webhookAuthenticatorCertRequester) sync(ctx context.Context, syncCtx factory.SyncContext) (progressing *bool, err error) {
+	key, secretComplete, shouldResubmit, err := c.checkKeyCertSecret(ctx, syncCtx.Recorder())
+	if err != nil {
+		return nil, err
+	}
+
+	if secretComplete {
+		return pbool(false), err
+	}
+
+	if !shouldResubmit {
+		var cert []byte
+		cert, shouldResubmit, err = c.handleCSR(key, syncCtx.Recorder())
+		if err != nil {
+			return nil, err
+		}
+
+		if cert != nil {
+			c.attachCertToSecret(ctx, cert)
+			// we've got the cert, stop progressing
+			return pbool(false), nil
+		}
+
+		if !shouldResubmit { // keep on waiting for the current CSR
+			return pbool(true), nil
+		}
+
+	}
+
+	// either we did not have a key or the CSR failed
+	if shouldResubmit {
+		if err := c.resubmitCSR(ctx, key); err != nil {
+			return nil, err
+		}
+		return pbool(true), nil
+	}
+
+	return pbool(false), nil
```

Comment:
it shouldn't under normal circumstances

Reply:
I don't think it will:
- at line 118 shouldResubmit==true => line 140 or 142  will return.
- at line 118 shouldResubmit==true => shouldResubmit is overwritten in line 120. If it is false now, line 132 return. If it is true now, line 140 or 142 return.